File: Jasmine/JasmineCompiler/Src/DFIR_Compiler/DFIR_Compiler/GetZeroIndegreeNodes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DFIR_Compiler
{
    public class GetZeroIndegreeNodes
    {
        public virtual LinkedList<INode> ZeroIndegreeNodesList(IDiagram iDiagram, bool isTopIDiagram)
        {
            LinkedList<INode> zeroindegreelist = new LinkedList<INode>();

            //Put the IConstant in current IDiagram into the zeroindegreelist
            foreach (IConstant temp in iDiagram.getIConstants())
            {
                if (temp.GetParentId() == iDiagram.GetNodeId())
                {
                    zeroindegreelist.AddLast(temp);
                }
            }

            //Put the INPUT IDataAccessor in current IDiagram into the zeroindegreelist
            foreach (IDataAccessor temp in iDiagram.getIDataAccessors())
            {
                if (temp.GetParentId() == iDiagram.GetNodeId() && temp.GetDirection() == Direction.INPUT)
                {
                    zeroindegreelist.AddLast(temp);
                }
            }

            //For the situation that the current IDiagram isn't the top IDiagram
            if (!isTopIDiagram)
            {
                INode n = INode.GetIdToINode()[iDiagram.GetParentId()];

                if (n.GetINodeType() == "ICaseStructure")
                {
                    ICaseStructure c = (ICaseStructure)n;
                    //Put the ICaseSelector into the zeroindegreelist
                    if (c.GetICaseSelector().IsConnectedIDiagram(iDiagram))
                    {
                        zeroindegreelist.AddLast(c.GetICaseSelector());
                    }
                    //Put the input ITunnel into the zeroindegreelist
                    foreach (ITunnel t in c.GetITunnels())
                    {
                        if (t.GetIsInput() && t.GetParentId() == iDiagram.GetParentId())
                        {
                            zeroindegreelist.AddLast(t);

                        }
                    }
                }

                else if (n.GetINodeType() == "IForLoop")
                {
                    IForLoop forloop = (IForLoop)n;

                    //Put the output connected ILoopIndex into the zeroindegreelist
                    if (forloop.GetILoopIndex().GetIsOutputConnected())
                    {
                        zeroindegreelist.AddLast(forloop.GetILoopIndex());
                    }
                    //Put the output connceted ILoopMax into the zeroindegreelist
                    if (forloop.GetILoopMax().GetIsOutputConnected())
                    {
                        zeroindegreelist.AddLast(forloop.GetILoopMax());
                    }
                    //Put the ILeftShiftRegister into the zeroindegreelist
                    foreach (ILeftShiftRegister ls in forloop.GetILeftShiftRegisters())
                    {
                        if (ls.GetParentId() == iDiagram.GetParentId())
                        {
                            zeroindegreelist.AddLast(ls);
                        }
                    }
                    //Put the input ITunnel into the zeroindegreelist
                    foreach (ITunnel t in forloop.GetITunnels())
                    {
                        if (t.GetIsInput() && t.GetParentId() == iDiagram.GetParentId())
                        {
                            zeroindegreelist.AddLast(t);
                        }
                    }

                }

            }

            return zeroindegreelist;
        }

    }
}

File: Wayne/WayneCompiler/src/Converter/GetZeroIndegreeNodes.java
package Converter;

import java.util.LinkedList;
import java.util.Queue;

import Dfir_representation.*;

//get the zero indegree nodes for the topological sort
public class GetZeroIndegreeNodes {
	public Queue<JNode> work(IDiagram iDiagram,Boolean isTopIDiagram){
		Queue<JNode> que = new LinkedList<JNode>();
		for(IConstant c : iDiagram.getIConstants()){
			if(c.getParentId() == iDiagram.getNodeId()){
				que.add(c);
			}
		}
		for(IDataAccessor c : iDiagram.getIDataAccessors()){
			if(c.getParentId() == iDiagram.getNodeId() && c.getDirection() == Direction.valueOf("INPUT")){
				que.add(c);
			}
		}
		for(IFeedbackOutputNode fo : iDiagram.getiFeedbackOutputNodes()){
			if(fo.getParentId() == iDiagram.getNodeId()){
				que.add(fo);
			}
		}
		//the list of zero indegree nodes should include iconstant, input IDataAccessor and IFeedbackOutputNode.
		
		//if it is not a top diagram, the list should include some input border nodes which are connected inside.
		if(!isTopIDiagram){
			JNode n = JNode.getIdToJNode().get(iDiagram.getParentId());
			if(n.getJNodeType() == "IForLoop"){
				IForLoop f = (IForLoop)n;
				if(f.getILoopIndex().getIsOutputConnected()) que.add(f.getILoopIndex());
				if(f.getILoopMax().getIsOutputConnected()) que.add(f.getILoopMax());
				for(ITunnel t : f.getITunnels()){
					if(t.getIsInput() && t.getParentId() == iDiagram.getParentId()){
						que.add(t);
					}
				}
				for(ILeftShiftRegister ls : f.getILeftShiftRegisters()){
					if(ls.getParentId() == iDiagram.getParentId()){
						que.add(ls);
					}
				}
			}
			/*else if(n.getJNodeType() == "IWhileLoop"){
				IWhileLoop f = (IWhileLoop)n;
				if(f.getILoopIndex().getIsOutputConnected()) que.add(f.getILoopIndex());
				for(ITunnel t : f.getITunnels()){
					if(t.getIsInput() && t.getParentId() == iDiagram.getParentId()){
						que.add(t);
					}
				}
				for(ILeftShiftRegister ls : f.getILeftShiftRegisters()){
					if(ls.getParentId() == iDiagram.getParentId()){
						que.add(ls);
					}
				}
			}*/
			else if(n.getJNodeType() == "ICaseStructure"){
				ICaseStructure c = (ICaseStructure) n;
				if(c.getiCaseSelector().getIsConnectedForIDiagram(iDiagram)){
					que.add(c.getiCaseSelector());
				}
				for(ITunnel t : c.getiTunnels()){
					if(t.getIsInput() && t.getParentId() == iDiagram.getParentId()){
						que.add(t);
						
					}
				}
			}
		}
		return que;
	}
}
